TimoCloud-Universal/src/main/java/cloud/timo/TimoCloud/base/managers/BaseInstanceManager.java

```diff
@@ -456,21 +449,42 @@ private void startProxy(BaseProxyObject proxy) {
         }
     }
 
-    private Integer getFreePort(int offset) {
+    private Integer getFreePortServer(BaseServerObject baseServerObject) throws ServerStartException {
+        Integer serverPort = getFreePortCommon(currentServerPort);
+
+        if (serverPort == null) {
+            TimoCloudBase.getInstance().severe("Error while starting server " + baseServerObject.getName() + ": No free port found. Please report this!");
+            throw new ServerStartException("No free port found");
+        }
+        serverPort++;
+        currentServerPort = serverPort;
+        if (currentServerPort.equals(maxServerPort)) currentServerPort = startPortServer;
+        return serverPort;
+    }
+
+    private Integer getFreePortProxy(BaseProxyObject baseProxyObject) throws ProxyStartException {
+        Integer proxyPort = getFreePortCommon(currentProxyPort);
+        if (proxyPort == null) {
+            TimoCloudBase.getInstance().severe("Error while starting proxy " + baseProxyObject.getName() + ": No free port found. Please report this!");
+            throw new ProxyStartException("No free port found");
+        }
+        proxyPort++;
+        currentProxyPort = proxyPort;
+        if (currentProxyPort.equals(maxProxyPort)) currentProxyPort = startPortProxy;
+        return proxyPort;
+    }
+
+    private Integer getFreePortCommon(int offset) {
```

Comment:
Change the signature so that the method accepts three parameters: `start`, `current` and `max`. Run the for loop `max-start` times and reset the counter to `start` when it reaches max. Change the max values to e.g. `41000` instead of `40999` so that you can simply check `if (current == max)`